### ISSUE TITLE
Upgrade the Mediator's ACA-Py instance to Release 0.8.1

### ISSUE THREAD

**Issue by swcurran** [bcgov/openshift-aries-mediator-service#25]
Please do a normal upgrade (through Dev, Test and then Prod) of the ACA-Py released used in the BC Wallet mediator to release 0.8.0.  This will pick up change [PR #2147](https://github.com/hyperledger/aries-cloudagent-python/pull/2147) - Fix: messages stuck in mediator.

~~Once done, please~~ During the upgrade, it is **required** to apply the upgrade outlined in the 0.8.0 Release notes to do with [PR #2116](https://github.com/hyperledger/aries-cloudagent-python/pull/2116).  Suggest confirming with an ACA-Py dev that the upgrade can be applied later versus as the update is made.  Here are the upgrade notes from the [0.8.0 Release Notes](https://github.com/hyperledger/aries-cloudagent-python/releases/tag/0.8.0) about this:

> **UPGRADE** [PR 2116](https://github.com/hyperledger/aries-cloudagent-python/pull/2116) - UPGRADE: Fix multi-use invitation performance
>
> The way that multiuse invitations in previous versions of ACA-Py caused performance to degrade over time. An update was made to add state into the tag names that eliminated the need to scan the tags when querying storage for the invitation.
>
> If you are using multiuse invitations in your existing (pre-0.8.0 deployment of ACA-Py, you can run an upgrade to apply this change. To run upgrade from previous versions, use the following command using the 0.8.0 version of ACA-Py, adding you wallet settings:
>
> `aca-py upgrade <other wallet config settings> --from-version=v0.7.5 --upgrade-config-path ./upgrade.yml`

**Comment by ianco** [comment]
The upgrade should be run pretty soon after the upgrade.  It just re-saves the connection records (in order to add a new tag)  - I'm not sure what is the impact if the new tags aren't available, possibly it won't be able to find the connection it's looking for,

**Comment by swcurran** [comment]
I'm not sure it matters, but better to be safe than sorry.  I think this search only happens in the midst of the invitation process, and the fix is to return only the invitation uses that are at that state.

**Comment by ianco** [comment]
The default path is setup here:  https://github.com/hyperledger/aries-cloudagent-python/blob/main/aries_cloudagent/commands/upgrade.py#L25

You should be able to run the upgrade without providing a `--upgrade-config-path`

**Comment by WadeBarnes** [comment]
```
1004460000@aries-mediator-agent-37-nsnw9-debug:~/.local/lib/python3.9/site-packages/aries_cloudagent/commands$ ls -1
__init__.py
__pycache__
help.py
provision.py
start.py
tests
upgrade.py
```

**Comment by ianco** [comment]
Hmmm that's an issue with the docker build I think ...

**Comment by ianco** [comment]
Do you have a way to manually copy the file to the mediator image?

**Comment by WadeBarnes** [comment]
The issue looks more like and issue with the `aca-py` python package.  The file is missing from the `bcgovimages/aries-cloudagent:py36-1.16-1_0.8.0` image too.  That one does a pip install of aca-py, while the one we're using from the aca-py repo builds the wheel and then installs it.

**Comment by WadeBarnes** [comment]
> Do you have a way to manually copy the file to the mediator image?

Yeah, just a pain.

**Comment by WadeBarnes** [comment]
Confirmed, the file is not included in either the source or wheel distribution on PyPi:
![image](https://user-images.githubusercontent.com/10504350/227262139-213df0ba-52d2-44ec-9816-756ad0e0ceb6.png)

**Comment by ianco** [comment]
I'll create an issue in the aca-py repo to make sure this file getting included in the release

**Comment by WadeBarnes** [comment]
Thanks

**Comment by WadeBarnes** [comment]
Running into issues with the upgrade.  Seems to be connectivity issues with the wallet database.  Investigating. 
```
1004460000@aries-mediator-agent-37-nsnw9-debug:/acapy-mediator$ $(echo aca-py upgrade \
  --wallet-type indy \
  --wallet-storage-type postgres_storage \
  --wallet-storage-config "$(eval echo \"${WALLET_STORAGE_CONFIGURATION}\")" \
  --wallet-storage-creds "$(eval echo \"${WALLET_STORAGE_CREDENTIALS}\")" \
  --from-version=v0.7.5 \
  --upgrade-config-path /tmp/default_version_upgrade_config.yml \
);
2023-03-23 16:16:03,870 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.actionmenu
2023-03-23 16:16:03,871 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.basicmessage
2023-03-23 16:16:03,871 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.connections
2023-03-23 16:16:03,962 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.coordinate_mediation
2023-03-23 16:16:03,963 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.didexchange
2023-03-23 16:16:03,963 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.discovery
2023-03-23 16:16:03,964 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.endorse_transaction
2023-03-23 16:16:03,965 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.introduction
2023-03-23 16:16:03,965 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.issue_credential
2023-03-23 16:16:04,170 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.notification
2023-03-23 16:16:04,171 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.out_of_band
2023-03-23 16:16:04,171 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.present_proof
2023-03-23 16:16:04,663 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.problem_report
2023-03-23 16:16:04,664 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.revocation_notification
2023-03-23 16:16:04,664 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.routing
2023-03-23 16:16:04,665 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.protocols.trustping
2023-03-23 16:16:04,666 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.holder
2023-03-23 16:16:05,963 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.ledger
2023-03-23 16:16:06,163 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.messaging.credential_definitions
2023-03-23 16:16:06,263 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.messaging.schemas
2023-03-23 16:16:06,268 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.messaging.jsonld
2023-03-23 16:16:06,271 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.revocation
2023-03-23 16:16:06,371 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.resolver
2023-03-23 16:16:06,462 aries_cloudagent.core.plugin_registry DEBUG Loaded module: aries_cloudagent.wallet
2023-03-23 16:16:07,065 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::core::startup?$'), processor <function on_startup_event at 0x7f9734cda550>
2023-03-23 16:16:07,065 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::core::shutdown?$'), processor <function on_shutdown_event at 0x7f9734cda5e0>
2023-03-23 16:16:07,669 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::REVOCATION::published.*'), processor <function on_revocation_published at 0x7f9734b31a60>
2023-03-23 16:16:07,670 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::REVOCATION::clear-pending.*'), processor <function on_pending_cleared at 0x7f9734b31f70>
2023-03-23 16:16:07,763 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::REVOCATION::published.*'), processor <function on_revocation_published at 0x7f9734adc280>
2023-03-23 16:16:07,763 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::REVOCATION::clear-pending.*'), processor <function on_pending_cleared at 0x7f9734adc9d0>
2023-03-23 16:16:07,766 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::CRED_DEF::(.*)?$'), processor <function on_cred_def_event at 0x7f9734f81b80>
2023-03-23 16:16:07,766 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::SCHEMA::(.*)?$'), processor <function on_schema_event at 0x7f9734f1a550>
2023-03-23 16:16:07,766 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::REVOCATION::REGISTRY_INIT.*'), processor <function on_revocation_registry_init_event at 0x7f9734e88040>
2023-03-23 16:16:07,767 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::REVOCATION::REGISTRY_ENDORSED.*'), processor <function on_revocation_registry_endorsed_event at 0x7f9734e88940>
2023-03-23 16:16:07,767 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::REVOCATION::SEND_ENTRY.*'), processor <function on_revocation_entry_event at 0x7f9734e880d0>
2023-03-23 16:16:07,768 aries_cloudagent.core.event_bus DEBUG Subscribed: topic re.compile('^acapy::ENDORSE_DID::(.*)?$'), processor <function on_register_nym_event at 0x7f9734eb6430>
2023-03-23 16:16:07,862 aries_cloudagent.core.profile INFO Create profile manager: indy
2023-03-23 16:16:07,864 aries_cloudagent.indy.sdk.wallet_plugin INFO Checking input postgres storage_config and storage_creds arguments
2023-03-23 16:16:07,864 aries_cloudagent.indy.sdk.wallet_plugin INFO Initializing postgres wallet
2023-03-23 16:16:07,962 aries_cloudagent.indy.sdk.wallet_plugin INFO Success, loaded postgres wallet storage
2023-03-23 16:16:07,962 indy.wallet DEBUG open_wallet: >>> config: '{"id": "aries-mediator-agent-wallet", "freshness_time": false, "storage_type": "postgres_storage", "storage_config": {"url": "aries-mediator-db:5432", "wallet_scheme": "DatabasePerWallet"}}', credentials: '{"key": "8mwXbYBRqlCerDqGjPN3AaiK2pvkekrD", "key_derivation_method": "ARGON2I_MOD", "storage_credentials": {"account": "VVNFUl9ZbzVI", "password": "aXVuN09YUGh1VUZ1WXIySg==", "admin_account": "postgres", "admin_password": "UldrMzlzbGdyaml6VGQ4cw=="}}'
2023-03-23 16:16:07,962 indy.wallet DEBUG open_wallet: Creating callback
2023-03-23 16:16:07,962 indy.libindy DEBUG create_cb: >>> cb_type: <class 'ctypes.CFUNCTYPE.<locals>.CFunctionType'>
2023-03-23 16:16:07,964 indy.libindy DEBUG create_cb: <<< res: <CFunctionType object at 0x7f9734b1df40>
2023-03-23 16:16:07,964 indy.libindy DEBUG do_call: >>> name: indy_open_wallet, args: (c_char_p(140287400351120), c_char_p(140287401309680), <CFunctionType object at 0x7f9734b1df40>)
2023-03-23 16:16:07,965 indy.libindy DEBUG _load_cdll: >>>
2023-03-23 16:16:07,965 indy.libindy DEBUG _load_cdll: Detected OS name: linux
2023-03-23 16:16:07,965 indy.libindy DEBUG _load_cdll: Resolved libindy name is: libindy.so
2023-03-23 16:16:07,965 indy.libindy DEBUG _load_cdll: <<< res: <CDLL 'libindy.so', handle 55e7d4849f30 at 0x7f9734aecc70>
2023-03-23 16:16:07,965 indy.libindy DEBUG set_logger: >>>
2023-03-23 16:16:07,965 indy.libindy DEBUG do_call_sync: >>> name: indy_set_logger, args: (None, None, <CFunctionType object at 0x7f9734b1dc40>, None)
2023-03-23 16:16:07,965 indy.libindy DEBUG do_call_sync: <<< 0
2023-03-23 16:16:07,966 indy.libindy DEBUG set_logger: <<<
2023-03-23 16:16:07,966 indy.libindy DEBUG do_call: Function indy_open_wallet returned err: 0
2023-03-23 16:16:07,966 indy.libindy DEBUG do_call: <<< <Future pending>

2023-03-23 16:18:19,329 indy.libindy DEBUG _get_error_details: >>>
2023-03-23 16:18:19,329 indy.libindy DEBUG _get_error_details: <<< error_details: {'backtrace': '', 'message': 'Error: Wallet not found\n  Caused by: Plugin returned error\n'}
2023-03-23 16:18:19,329 indy.libindy DEBUG _indy_callback: >>> command_handle: 0, err , args: (0,)
2023-03-23 16:18:19,329 indy.libindy DEBUG _indy_callback: <<<
2023-03-23 16:18:19,329 indy.libindy DEBUG _indy_loop_callback: >>> command_handle: 0, err , args: (0,)
2023-03-23 16:18:19,330 indy.libindy WARNING _indy_loop_callback: Function returned error 
2023-03-23 16:18:19,330 indy.libindy DEBUG _indy_loop_callback <<<
Traceback (most recent call last):
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/indy/sdk/wallet_setup.py", line 167, in open_wallet
    handle = await indy.wallet.open_wallet(
  File "/usr/local/lib/python3.9/site-packages/indy/wallet.py", line 124, in open_wallet
    res = await do_call('indy_open_wallet',
  File "/usr/local/lib/python3.9/asyncio/futures.py", line 284, in __await__
    yield self  # This tells Task to wait for completion.
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 328, in __wakeup
    future.result()
  File "/usr/local/lib/python3.9/asyncio/futures.py", line 201, in result
    raise self._exception
indy.error.WalletNotFoundError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/commands/upgrade.py", line 107, in upgrade
    root_profile, public_did = await wallet_config(context)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/config/wallet.py", line 54, in wallet_config
    yield self  # This tells Task to wait for completion.
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 328, in __wakeup
    future.result()
  File "/usr/local/lib/python3.9/asyncio/futures.py", line 201, in result
    raise self._exception
aries_cloudagent.core.error.ProfileNotFoundError: Wallet 'aries-mediator-agent-wallet' not found: Error: Wallet not found
  Caused by: Plugin returned error


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/__main__.py", line 74, in <module>
    main(sys.argv)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/__main__.py", line 71, in main
    run(args)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/__main__.py", line 64, in run
    run_command(command, args)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/commands/__init__.py", line 38, in run_command
    module.execute(argv)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/commands/upgrade.py", line 241, in execute
    loop.run_until_complete(upgrade(settings))
  File "/home/indy/.local/lib/python3.9/site-packages/nest_asyncio.py", line 90, in run_until_complete
    return f.result()
  File "/usr/local/lib/python3.9/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 258, in __step
    result = coro.throw(exc)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/commands/upgrade.py", line 218, in upgrade
    raise UpgradeError(f"Error during upgrade: {e}")
aries_cloudagent.commands.upgrade.UpgradeError: Error during upgrade: Wallet 'aries-mediator-agent-wallet' not found: Error: Wallet not found
  Caused by: Plugin returned error
  ```

**Comment by WadeBarnes** [comment]
Error is now:
```
Traceback (most recent call last):
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/commands/upgrade.py", line 147, in upgrade
    raise UpgradeError(
aries_cloudagent.commands.upgrade.UpgradeError: No upgrade configuration found for v0.7.5

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/__main__.py", line 74, in <module>
    main(sys.argv)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/__main__.py", line 71, in main
    run(args)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/__main__.py", line 64, in run
    run_command(command, args)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/commands/__init__.py", line 38, in run_command
    module.execute(argv)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/commands/upgrade.py", line 241, in execute
    loop.run_until_complete(upgrade(settings))
  File "/home/indy/.local/lib/python3.9/site-packages/nest_asyncio.py", line 90, in run_until_complete
    return f.result()
  File "/usr/local/lib/python3.9/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 256, in __step
    result = coro.send(None)
  File "/home/indy/.local/lib/python3.9/site-packages/aries_cloudagent/commands/upgrade.py", line 218, in upgrade
    raise UpgradeError(f"Error during upgrade: {e}")
aries_cloudagent.commands.upgrade.UpgradeError: Error during upgrade: No upgrade configuration found for v0.7.5
2023-03-23 16:24:22,363 aries_cloudagent.indy.sdk.profile DEBUG Profile finalizer called; closing wallet
```

Looks like you can only specify versions that are explicitly listed in the upgrade configuration yaml.

**Comment by swcurran** [comment]
Any notes about an upgrade to the documentation on upgrade would be appreciated.  Current “docs” are [here in the Changelog for the 0.8.0 release](https://github.com/hyperledger/aries-cloudagent-python/blob/main/CHANGELOG.md#breaking-changes-and-upgrades).  I don’t see any general documentation on “Upgrade”, but I didn’t look too hard.

It sounds like we need to do a new 0.8.1 immediately with the missing file.  Please let me know what needs to be in that file.  

Please put details in: https://github.com/hyperledger/aries-cloudagent-python/issues/2178

Or open a PR to add the necessary documents that would have helped you.

Thanks.

**Comment by WadeBarnes** [comment]
> Any notes about an upgrade to the documentation on upgrade would be appreciated. Current “docs” are [here in the Changelog for the 0.8.0 release](https://github.com/hyperledger/aries-cloudagent-python/blob/main/CHANGELOG.md#breaking-changes-and-upgrades). I don’t see any general documentation on “Upgrade”, but I didn’t look too hard.

The only tripping points were the `default_version_upgrade_config.yml` not existing in the release, knowing you could exclude `--upgrade-config-path` if it existed, having to specify a version for `--from-version` that exists in the file, and having to specify `--from-version=v0.8.0` when upgrading to `v0.8.0`.

**Comment by WadeBarnes** [comment]
I've run into a problem upgrading the `test` environment.  When running the upgrade I get the following output, which indicates the upgrade will not be run.
```
1004440000@aries-mediator-agent-24-v626k-debug:/acapy-mediator$ $(echo aca-py upgrade \
  --wallet-type indy \
  --wallet-storage-type postgres_storage \
  --wallet-storage-config "$(eval echo \"${WALLET_STORAGE_CONFIGURATION}\")" \
  --wallet-storage-creds "$(eval echo \"${WALLET_STORAGE_CREDENTIALS}\")" \
  --from-version=v0.7.5 \
  --upgrade-config-path /tmp/default_version_upgrade_config.yml \
);
2023-03-24 13:33:29,325 aries_cloudagent.core.profile INFO Create profile manager: indy
2023-03-24 13:33:29,326 aries_cloudagent.indy.sdk.wallet_plugin INFO Checking input postgres storage_config and storage_creds arguments
2023-03-24 13:33:29,327 aries_cloudagent.indy.sdk.wallet_plugin INFO Initializing postgres wallet
2023-03-24 13:33:29,332 aries_cloudagent.indy.sdk.wallet_plugin INFO Success, loaded postgres wallet storage
2023-03-24 13:33:30,259 aries_cloudagent.indy.sdk.wallet_setup INFO Creating master secret...
2023-03-24 13:33:30,261 indy.libindy WARNING _indy_loop_callback: Function returned error
2023-03-24 13:33:30,261 aries_cloudagent.indy.sdk.wallet_setup INFO Master secret already exists
version v0.8.0 found in storage, --from-version will be ignored.
Version v0.8.0 to upgrade from and current version to upgrade to v0.8.0 are same.
```

@ianco, Any thoughts?

**Comment by swcurran** [comment]
The error message is the same as we saw with the test failure, right?   Weird that this was not hit in Dev — such fun.  Is it because we were running 0.8.0-rc1 there?  It’s considered the same as 0.8.0?

**Comment by ianco** [comment]
We need to relax the rules on the upgrade.  This upgrade is non-destructive (it just re-saves the connection records to update the tags) so shouldn't be an issue if it's re-run.  However it's weird that the storage version was changed to `0.8.0` before the upgrade was run.  Also the upgrade shouldn't enforce the exact "from" version to be in the `upgrade.yml` file.

Not sure about the `0.8.0` vs `0.8.0-rc1` issue why wouldn't the storage version be `0.8.0-rc1` if that's what the aca-py version is?

**Comment by WadeBarnes** [comment]
Both `dev` and `test` were running `0.8.0-rc0` previously.  I also noticed after looking at the upgrade code that none of the expected log messages from the upgrade are contained in the log output from when I ran the upgrade on `dev`, so I'm not convinced the upgrade ran successfully there either.

**Comment by ianco** [comment]
I suggest:
- add a `--force` flag to the `upgrade` command to force the upgrade to run, regardless of the version check
- add the `version` info to the `/status/config` output so we can see what the version is
- when we do a `rc` release, make sure the version in the `version.py` reflects the `rc` number (it looks like `0.8.0-rc0` was recorded as `0.8.0` in the wallet storage)  (or wherever this is stored make sure the `rc` info is included in the version)
- add an unencrypted tag to the `version` record in the wallet so we can inspect the db to figure out what is the version (and which record it is)

Maybe also look at the logic that determines whether the upgrade should run - it looks like it expects the exact `from` version to be in the `upgrade.yml` file, which it isn't (we have only been including the version info where there is an actual upgrade in the release).

**Comment by swcurran** [comment]
@WadeBarnes — can you try running the command with `--from-version=v0.7.2`, since that is the last one in the file?

**Comment by swcurran** [comment]
Can you tell from the code (a) what goes into the storage, and (b) if the comparison that is being used would detect 0.8.0-rc0 as later than 0.8.0?  

I’m certain that the value is `version.py` is accurate for each release — that’s the only thing we change, it is verified, and we can confirm that for every release.  I can’t see that as the issue.

If a value is recorded in secure storage, we should be able to see that, don’t you think?  We don’t really need an unencrypted tag if we assume that no one is sneaking around updating it another way.

**Comment by ianco** [comment]
The version record in the wallet is updated here, in the upgrade command:

https://github.com/hyperledger/aries-cloudagent-python/blob/main/aries_cloudagent/commands/upgrade.py#L202

It looks like it just saves the version from `version.py`, so it should include the `rc` part of the version.

There's a helper method here for updating the version in the wallet but it's not called anywhere:

https://github.com/hyperledger/aries-cloudagent-python/blob/main/aries_cloudagent/config/wallet.py#L142

... and it looks like the Conductor will bail on startup if the version doesn't match:

https://github.com/hyperledger/aries-cloudagent-python/blob/main/aries_cloudagent/core/conductor.py#L313

(... unless there is no version record in the wallet, in which case aca-py will startup happily)

**Comment by ianco** [comment]
> If a value is recorded in secure storage, we should be able to see that, don’t you think?

All the contents of the wallet are encrypted so we can't inspect anything ...  we can't connect to the wallet database directly and see anything, and the version info stored in the wallet isn't exposed anywhere via the api

**Comment by swcurran** [comment]
Just realized @WadeBarnes that you used `—from-version 0.8.0` on Dev.  That doesn’t really make sense, as the from version should be less than that, right?

The “old” (from) version either comes from Storage, or if not found, comes from the `—from-version` arg, and the “to” version is set in `version.py`.

It also looks like the version is never added to storage unless an upgrade has been run at least once — the add record is only (AFAICS) in the upgrade command.

And I certainly can’t see that the value `version.py` is ever altered — e.g. the `-rc0` dropped.

Frustrating...

**Comment by swcurran** [comment]
> > If a value is recorded in secure storage, we should be able to see that, don’t you think?
> 
> All the contents of the wallet are encrypted so we can't inspect anything ... we can't connect to the wallet database directly and see anything, and the version info stored in the wallet isn't exposed anywhere via the api

I mean — we should be able to see where the data is updated in the code such that we don’t have to see the literal value.

**Comment by ianco** [comment]
> I mean — we should be able to see where the data is updated in the code such that we don’t have to see the literal value.

Sure but that doesn't tell us what version is currently running, for example what is the storage version in the production mediator right now?  We can't tell ...

**Comment by swcurran** [comment]
The only way to get the version into storage is to run the upgrade, and since it is pretty unlikely an upgrade has been run on this instance before, there should be no version in storage.  How is that routine returning a value?

This seems like the same issue we had with the integration test just before release.  What did you do to fix that @ianco? Is it possible we’re seeing the same thing?

**Comment by WadeBarnes** [comment]
No, having to specify an explicit version contained in a file you can't see easily does not make sense.  I mentioned that [here](https://github.com/bcgov/openshift-aries-mediator-service/issues/25#issuecomment-1481622404).  I had originally specified `v0.7.5` as recommended in the notes provided but the result was [this](https://github.com/bcgov/openshift-aries-mediator-service/issues/25#issuecomment-1481505741).

I'm thinking it would be way better if `aca-py` managed it's own upgrade lifecycle on start-up rather than having to run the upgrade out-of-band.  It could warn you if you are running a different version than specified in storage and halt until you confirm, somehow (like an environment variable to flag you want to upgrade/downgrade), that what you're doing is intentional.

I'm a little concerned now about the Conductor bailing on startup when the version in storage does not match the running version.  Our automated upgrades have been working to this point because we've never done and upgrade, so there is no version record in storage.  Now that we've done an upgrade the version is in storage so all future automated deployments to our (say `dev`) environments will fail because the Conductor will bail whenever we try to deploy a new version.  We'll have to update the deployment configs to include an upgrade step.  The question then becomes what happens if we need to roll back to a previous version due to issue discovered in a new vesion?

**Comment by swcurran** [comment]
I think we need to be working together on this on a call. I’ll add @shaangill025. @ianco — not sure if you are working today or not, so you may not be able to join.

Completely agree with you about bailing on startup if the upgrade is not done.  That seems like a bad idea.  But at the same time, if the upgrade is executed by 5 instances of ACA-Py on the same storage, what will happen…?

For now, we need to understand why the query of secure storage is saying we are on version 0.8.0, when an upgrade has never been run...

**Comment by WadeBarnes** [comment]
The version likely got in there when I ran the upgrade the first time.  What I posted was not actually the first run.  I had originally run it with `--from-version=v0.8.0` like I did in `dev` which seemed to work, but now on further examination I'm not confident it actually did as noted earlier.

The result at that point was something like this simulated log output:
```
1004440000@aries-mediator-agent-24-v626k-debug:/acapy-mediator$ $(echo aca-py upgrade \
  --wallet-type indy \
  --wallet-storage-type postgres_storage \
  --wallet-storage-config "$(eval echo \"${WALLET_STORAGE_CONFIGURATION}\")" \
  --wallet-storage-creds "$(eval echo \"${WALLET_STORAGE_CREDENTIALS}\")" \
  --from-version=v0.8.0 \
  --upgrade-config-path /tmp/default_version_upgrade_config.yml \
);
2023-03-24 13:33:29,325 aries_cloudagent.core.profile INFO Create profile manager: indy
2023-03-24 13:33:29,326 aries_cloudagent.indy.sdk.wallet_plugin INFO Checking input postgres storage_config and storage_creds arguments
2023-03-24 13:33:29,327 aries_cloudagent.indy.sdk.wallet_plugin INFO Initializing postgres wallet
2023-03-24 13:33:29,332 aries_cloudagent.indy.sdk.wallet_plugin INFO Success, loaded postgres wallet storage
2023-03-24 13:33:30,259 aries_cloudagent.indy.sdk.wallet_setup INFO Creating master secret...
2023-03-24 13:33:30,261 indy.libindy WARNING _indy_loop_callback: Function returned error
2023-03-24 13:33:30,261 aries_cloudagent.indy.sdk.wallet_setup INFO Master secret already exists
Version v0.8.0 to upgrade from and current version to upgrade to v0.8.0 are same.
```

**Comment by swcurran** [comment]
OK — that’s good to know.  And that command would have happily skipped the update and updated the storage version.  Phew…we know how we got here.  Now we need to decide what we need to do, and what to do for the long term.

I think we switch the logic on the check such that the `—from-version` overrides the storage version (instead of the other way round).

The bigger question that perhaps we don’t have to decide today is how we fix the automated deployment issue that Wade highlights above.  That needs safe automation.

Meeting set for this morning.

**Comment by shaangill025** [comment]
After taking a look at the code for `upgrade` function, I think the config file for `v8.0.0` release [Assuming ConnRecord for all previous 0.7.* versions need to be resaved] should be something like this:
```
v0.7.5:
  resave_records:
    base_record_path:
      - "aries_cloudagent.connections.models.conn_record.ConnRecord"
  update_existing_records: false
v0.7.4:
  resave_records:
    base_record_path:
      - "aries_cloudagent.connections.models.conn_record.ConnRecord"
  update_existing_records: false
v0.7.3:
  resave_records:
    base_record_path:
      - "aries_cloudagent.connections.models.conn_record.ConnRecord"
  update_existing_records: false
v0.7.2:
  resave_records:
    base_record_path:
      - "aries_cloudagent.connections.models.conn_record.ConnRecord"
  update_existing_records: false
v0.7.1:
  resave_records:
    base_record_path:
      - "aries_cloudagent.connections.models.conn_record.ConnRecord"
  update_existing_records: false
v0.7.0:
  resave_records:
    base_record_path:
      - "aries_cloudagent.connections.models.conn_record.ConnRecord"
  update_existing_records: false
v0.6.0:
  update_existing_records: false
```

If there was no version record in storage initially then running with `--from-version=0.7.*` should have worked. But as the version record has already been set as `0.8.0` without the ConnRecord being resaved, I think we implement `--force-upgrade` like @ianco suggested which ignores the version record from storage and instead uses the `--from-version` to rerun the upgrade process.

**Comment by swcurran** [comment]
Issue: How to force an upgrade:
 
 - Decision is to have both `--force-upgrade` and `--from-version`
 
 Issue: What if the from-version is not in the `upgrade` yaml?  It would be nice to have it fallback to the one previous to the `--from-version` or in secure storage.
 
 - Decision: Make the code smarter and have it find the first version equal to or greater than the `from` version (from args or storage) and apply all the upgrades including and to the current version.  This becomes the literal from to use.
 - Further -- if there is more than one identical update -- run it only once. See the example above -- all of the updates are to `resave-ConnRec` so it really only needs to be done once.

Issue: Once a version is in storage, cannot move to a newer version without running the `upgrade` command.

- Decision: Change to only stop if there is an upgrade required between versions -- and if not, update the storage version.

Issue: On a new deployment, there is no version put into storage.

- Decision: Add to the create storage the additional of the version record.

Issue: No version in storage in existing deployments.

- Decision: Recommend that the deployment have an `upgrade --from-version...` command.

Issue: No warning currently produced when version in storage is not found.

- Decision: Add the warning, with details about what to do.

Issue: Can we run an upgrade automatically when deploying a new version. Current behaviour is to abort the startup.

- Option: Auto-upgrade.
   - Risk: Multiple instances try to run the upgrade could cause unexpected side effects.
      - Same as the next problem -- must be run offline
      - ~~Need a semaphore or equivalent if the upgrade must be run once and only once.~~
   - Risk: Some updates MAY require running the upgrade offline.
      - Flag in YML file to say MUST be run as an explicit upgrade.

Issue: The current expectation is that in the YML, the version for the change is the "from" -- e.g. when upgrading from a designated release.

- Decision: The version associated with an upgrade command should be the version in when the upgrade should be run. E.g. Any version being upgraded to that version must run the upgrade steps.  So:
   - If the entry is `v0.8.0: Update_Connections` -- that command should be run from every pre-0.8.0 version when upgrade to 0.8.0 or beyond.  Further, if going from 0.7.5 to 0.9.0, they would run the `0.8.0` upgrade.

**Comment by WadeBarnes** [comment]
All mediator instances have been successfully upgrade to `aca-py v0.8.1`.  Related PR; https://github.com/bcgov/openshift-aries-mediator-service/pull/28
